astropy/units/quantity_helper/function_helpers.py

```diff
@@ -514,68 +515,128 @@ def _block(arrays, max_depth, result_ndim, depth=0):
-        k: v
-        for k, v in (("stop", stop), ("start", start), ("step", step))
-        if v is not None
-    }
+
+    def parse_start_stop_step(*, start_or_stop, _stop, _step):
+        # handle the perilous task of disentangling original arguments
+        # this isn't trivial because start_or_stop may actually bind to two
+        # different inputs, as the name suggests
+        # we (ab)use structural pattern matching here to bind output variables
+        # (start, stop, step), so n additional logic is actually need after a
+        # match is found.
+        match (start_or_stop, _stop, _step):
+            case (stop, None as start, ArangeDefaults.STEP as step):
+                pass
+            case (start, stop, ArangeDefaults.STEP as step):
+                pass
+            case (start, stop, step):
+                pass
+
+        return start, stop, step
+
+    start, stop, step = parse_start_stop_step(
+        start_or_stop=start_or_stop, _stop=stop, _step=step
+    )
     out_unit = getattr(stop, "unit", UNIT_FROM_LIKE_ARG)
-    if out_unit is UNIT_FROM_LIKE_ARG:
-        if hasattr(start, "unit") or hasattr(step, "unit"):
-            raise TypeError(
-                "stop without a unit cannot be combined with start or step with a unit."
-            )
-        kwargs.update(qty_kwargs)
-    else:
-        # Convert possible start, step to stop units.
-        new_values, _ = _quantities2arrays(*qty_kwargs.values())
-        kwargs.update(zip(qty_kwargs.keys(), new_values))
+
+    if out_unit is UNIT_FROM_LIKE_ARG and (
+        hasattr(start, "unit") or hasattr(step, "unit")
+    ):
+        raise TypeError(
+            "stop without a unit cannot be combined with start or step with a unit."
+        )
+
+    def wrapup_arguments(*, start, stop, step):
+        # do the reverse operation than parse_start_stop_step
+        # this is needed because start_or_stop *must* be passed as positional
+        # (starting in numpy 2.4)
+
+        # purely defensive programming: if this ever raises,
+        # it means a mistake was already made at this point,
+        # possibly upstream of this function.
+        assert stop is not None, "Please report this."
+
+        match start, stop, step:
+            case (None, _, ArangeDefaults.STEP):
+                qty_args = (stop,)
+                kwargs = {}
+            case (_, _, ArangeDefaults.STEP):
+                qty_args = (start, stop)
+                kwargs = {}
+            case (None, _, _):
+                qty_args = (stop,)
+                kwargs = {"step": step}
+            case _:
+                qty_args = (start, stop)
+                kwargs = {"step": step}
+
+        # reverse positional arguments so `stop`` always comes first
+        # this is done to ensure that the arrays are first converted to the
+        # expected unit, which we guarantee should be stop's
+        args_rev, out_unit_loc = _quantities2arrays(*reversed(qty_args))
```

Comment:
overall, I'm not sure using `_quantities2arrays` is a good idea here, but it's the easiest way to ensure backward compat with the previous implementation, so I'll give it a rest for now.